[Proposal] OpenAPI for S3 multipart upload (#7197)
treeverse/lakeFS: design/open/openapi-multipart-upload.md

# OpenAPI for S3 Like Multipart Upload Proposal

## Introduction

This proposal outlines the development of Open API endpoints to enable multipart upload capabilities similar to Amazon S3 multipart. The intention is to enhance the data uploading process for large files, ensuring efficient and reliable data transfer.

## Goal

- Implementing multi-part upload APIs with support for parallel upload optimizes data transfer for large files over varying network conditions using lakectl.

## Non-goals

- Support all underlying storage types as we can experiment with the one that currently support this feature.
- SDK to wrap the new API and support this capability.

## Features and Functionalities

> In order to minimaize the scope of this feature and understand the benefits before supporting all blockstores, this feature will be scoped to S3 blockstore and only if presigned capability is enabled.

Reviewer: ```suggestion
To minimize the scope of this feature and understand the benefits before supporting all blockstores, this feature will be scoped to S3 blockstore and only if presigned capability is enabled.
```

Support for Large Files: The API will handle files of substantial sizes, ensuring that large datasets can be uploaded without issues. Minimum part size will be 5M, except for the last block.

Data Integrity Checks: Each uploaded part will require the client to store the ETag as done today for presigned URL upload. The value of the ETag will be provided when the upload of all the parts is completed.

Pause and Resume: Clients can pause the upload and resume later, leveraging the upload ID to reinitiate the process. (TDB need to specify how long we will keep and if cleanup will be done)

Completion of Upload: Once all parts are uploaded, the client will send a request to complete the upload, address the file in lakeFS as a single file. Upload compete request will include the ETag for check part, in case of mismatch or fail to provide bad request code will return.

Cancelation: In order to cancel partial upload of multipart request, the client will require to call cancellation with the upload ID provided.

### Multipart upload capability using OpenAPI specification

**Paths for multipart upload operations**

```yaml

>   /repositories/{repository}/branches/{branch}/staging/multipart:

Author: Good point!
- Didn't want to open another interface to upload data through lakefs
- The requests to create/complete/abort multipart works on upload ID that represents the operation. The presined parts currently returned on CreateMultipartUpload operations only if you specify the number of parts you like to get and it is optional. It means that if we like in the future to open upload part data through lakeFS we can still provide this API.
- I will update the spec to rename `parts` to `presigned_parts` to indicate this number represents the number of urls the call will return.

>   /repositories/{repository}/branches/{branch}/staging/multipart:

Reviewer 2: I don't see the alternative as uploading through lakeFS. I see the non-pre-signed version of this using local credentials on the client, i.e., the client uses the AWS SDK to upload directly to an S3, according to a physical address returned from this endpoint.  

But I'm also ok with not doing this now and seeing if there's a need there.

>   /repositories/{repository}/branches/{branch}/staging/multipart:

Author: Both ways - using presign and non-presigned URLs require local credentials to work.
Supporting multipart upload using the API while data transfer goes though lakeFS is possible, but I would like to avoid it and just provide more ways to enable the client direct access to the specific object using the underlying object store.

        required: true
        schema:
          type: string
      - in: path
        name: branch
        required: true
        schema:
          type: string
      - in: query
        name: path
        description: relative to the branch
        required: true
        schema:
          type: string
      - in: query
        name: parts
        description: number of parts required to upload
        schema:
          type: integer
    post:
      tags:
        - staging
      operationId: createMultipartUpload
      summary: Initiate a multipart upload
      description: Initiates a multipart upload and returns an upload ID with presigned URLs for each part. Part numbers starts with 1. Each part minimum size is 5M.
      responses:
        200:
          description: Multipart upload initiated
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/MultipartUpload"
        # ... other error responses

  /repositories/{repository}/branches/{branch}/staging/multipart/{uploadId}:
      parameters:
      - in: path
        name: repository
        required: true
        schema:
          type: string

>       - in: path

Author: Added - thanks!

          items:
            $ref: "#/components/schemas/UploadPart"
      required:
        - parts
```

### Support and discover

Presign support is a capability lakectl discover before switching to use presign for upload or download from lakeFS.
The multipart upload support will be part of the storage capability add `pre_sign_support_parts` optional field that when set to `true` the user can perform multipart upload using the new API.

```yaml
    StorageConfig:
      type: object
      required:
        - blockstore_type
        - blockstore_namespace_example
        - blockstore_namespace_ValidityRegex
        - pre_sign_support
        - pre_sign_support_ui
        - import_support
        - import_validity_regex
      properties:
        blockstore_type:
          type: string
        blockstore_namespace_example:
          type: string
        blockstore_namespace_ValidityRegex:
          type: string
        default_namespace_prefix:
          type: string
        pre_sign_support:
          type: boolean
        pre_sign_support_ui:
          type: boolean
        import_support:
          type: boolean
        import_validity_regex:
          type: string

>         pre_sign_support_parts:

Reviewer 2: If the presigned requirement is a limitation for the purposes of an MVP, then it should be a capability of the feature, as it is in other endpoints.

I'd suggest naming the config property something like `multipart_upload_support`. Especially since we already have `pre_sign_support`. The current limitation is that you can't set `multipart_upload_support` to `true` if `pre_sign_support` is set to `false`. We can "enforce" it this way.  

If we decide to allow multipart uploads without the presign requirement, we can remove this check.

Author: Update, the client will have to check both capabilities presigned and mutlipart supprort before using the API. The curent implementation is not return true for multipart upload support if presigned is off, as it is the only configuration that works.

          type: boolean
```

## Limitations

> - **S3 block adapter exclusive:** Uploading files in multiple parts (multipart upload) is only available when using the S3 block adapter. This feature isn't currently supported with other storage options.

Reviewer 2: Maybe mention here that multipart uploads are available through the S3 gateway to other block adapters so readers won't get the wrong idea that multipart uploads aren't supported at all for non-S3 users.

- **Part size and count restrictions:** There are limits on how you can split your file for upload:
    - **Maximum parts:** You can split your file into a maximum of 10000 parts.
    - **Minimum part size:** Each part must be at least 5MB in size. This is a temporary constraint and isn't currently configurable or discoverable. It will become an option when additional storage options are supported.
- **Initiating the upload:** When starting a multipart upload, you'll need to specify the total number of parts in your request. It reduce the requests for each part presigned URL when the client already knows the size.
- **Presigning part:** Request for a presigned URL of a specific part number will not be supported. This will block unknown size upload using this API.

> - **Limited support:** Request for list upload parts will be be provided at this point.

Reviewer 2: ```suggestion
- **Limited support:** Request for list upload parts will not be provided at this point.
```

# OpenAPI for S3 Like Multipart Upload Proposal

## Introduction

This proposal outlines the development of Open API endpoints to enable multipart upload capabilities similar to Amazon S3 multipart. The intention is to enhance the data uploading process for large files, ensuring efficient and reliable data transfer.

## Goal

- Implementing multi-part upload APIs with support for parallel upload optimizes data transfer for large files over varying network conditions using lakectl.

## Non-goals

- Support all underlying storage types as we can experiment with the one that currently support this feature.
- SDK to wrap the new API and support this capability.

## Features and Functionalities

> In order to minimaize the scope of this feature and understand the benefits before supporting all blockstores, this feature will be scoped to S3 blockstore and only if presigned capability is enabled.

Author: Looks I wanted to write it down as the last limitation but I forgot to complete the sentence :) thanks!

> Initiation of Upload: The API will allow clients to initiate a multipart upload session, assigning a unique upload ID for subsequent operations. The client will require pass the number of parts it requires and the call will provide a set of URLs to upload each part.

Author: Agree, any streaming or working with unknown size is good for multipart or at least working with "upload manager" that will perform multipart or single object upload.

In order to limit the scope of the implementation, it is a limitation as I didn't want the client to call lakeFS for each part.
Working with local file and known in advance the size we like to upload - optional list as part of the create to start multipart will handle it on a single request.
The suggested API doesn't require the server to return any presigned part so we can include this capability later without breaking the API.

> Initiation of Upload: The API will allow clients to initiate a multipart upload session, assigning a unique upload ID for subsequent operations. The client will require pass the number of parts it requires and the call will provide a set of URLs to upload each part.

Reviewer: OK, so perhaps document next possible step -- it isn't clear?

For instance somewhere...
```md
### Next steps

None of the returned URLs has to be used, it is fine to ask for more than are needed.  In future we may add an _additional_ API call to URLs for uploading more parts.  This will allow more "streaming" uses, for instance as parallels to how Hadoop S3A uses the S3 MPU API.  The current API is compatible with such an addition.
```

Uploading Parts: Clients can upload individual parts of the file in parallel or in sequence. Each part will be based on the presigned URL provided by the initial call.

> Support for Large Files: The API will handle files of substantial sizes, ensuring that large datasets can be uploaded without issues. Minimum part size will be 5M.

Author: I'll update that each part (exclude the last one) minimum size depends on the underlying blockstore implementation.

Data Integrity Checks: Each uploaded part will require the client to store the ETag as done today for presigned URL upload. The value of the ETag will be provided when the upload of all the parts is completed.

Pause and Resume: Clients can pause the upload and resume later, leveraging the upload ID to reinitiate the process. (TDB need to specify how long we will keep and if cleanup will be done)

Completion of Upload: Once all parts are uploaded, the client will send a request to complete the upload, address the file in lakeFS as a single file. Upload compete request will include the ETag for check part, in case of mismatch or fail to provide bad request code will return.

Cancelation: In order to cancel partial upload of multipart request, the client will require to call cancellation with the upload ID provided.

### Multipart upload capability using OpenAPI specification

**Paths for multipart upload operations**

> ```yaml=

Author: Thanks - my markdown editor use `yaml=` to render yaml + line numbers

        description: relative to the branch
        required: true
        schema:
          type: string
    put:
      tags:
        - staging
      operationId: completeMultipartUpload
      summary: Complete a multipart upload
      description: Completes a multipart upload by assembling the uploaded parts.
      requestBody:
        content:
          application/json:
            schema:
              $ref: "#/components/schemas/CompleteMultipartUpload"
      responses:
        200:
          description: Multipart upload completed
          content:
            application/json:
              schema:
                $ref: "#/components/schemas/ObjectStats"
        # ... other error responses

    delete:
      tags:
        - staging
      operationId: abortMultipartUpload
      summary: Abort a multipart upload
      description: Aborts a multipart upload.
      responses:
        204:
          description: Multipart upload aborted

```

**Schemas used by operations**

```yaml=
components:
  schemas:
    MultipartUpload:
      type: object
      properties:
        upload_id:
          type: string
        locations:
          type: array
          items:

>             $ref: "#/components/schemas/StagingLocation"

Reviewer: Obviously if we don't require the client to supply the number of parts in advance, then we cannot do this.  We will need another API call to get another presigned URL for an upload ID.

Author: True, in the current spec the number of `parts` in the request is optional. But will be required by the server.
And the server response include StagingLocations as optional in the case we will like to support dynamically request for part presigned URL(s).

      required:
        - upload_id
        
    UploadPart:
      type: object
      properties:
        part_number:
          type: integer
          minimum: 1

>           maximum: 1000

Reviewer: S3 allows [10_000 parts](https://docs.aws.amazon.com/AmazonS3/latest/userguide/qfacts.html).  In any case I do not believe that this limitation should be in the OpenAPI spec: if I later decide to change it on lakeFS, old clients should be able to use this higher maximum value.

        etag:
          type: string
      required:
        - part_number
        - etag
        
    CompleteMultipartUpload:
      type: object
      properties:
        parts:
          type: array
          items:

>             $ref: "#/components/schemas/UploadPart"

Reviewer: Perhaps worth mentioning in the description that the object is created of these parts in the specified order -- **NOT** in the order in which the objects were uploaded, or in the order that `createMultipartUpload` returned the parts.

Author: Good point - the parts should be sorted by part number.

Reviewer: Amusingly, the [S3 API](https://docs.aws.amazon.com/AmazonS3/latest/API/API_CompleteMultipartUpload.html) does not document such a requirement AFAICT?  But this is fine.

Author: Yea, you learn from the hard way.

```
Error Code: InvalidPartOrder

Description: The list of parts was not in ascending order. The parts list must be specified in order by part number.
HTTP Status Code: 400 Bad Request
```

Documented in https://docs.aws.amazon.com/cli/latest/reference/s3api/complete-multipart-upload.html

        pre_sign_support_ui:
          type: boolean
        import_support:
          type: boolean
        import_validity_regex:
          type: string
        pre_sign_support_parts:
          type: boolean
```

## Limitations

- **S3 block adapter exclusive:** Uploading files in multiple parts (multipart upload) is only available when using the S3 block adapter. This feature isn't currently supported with other storage options.
- **Part size and count restrictions:** There are limits on how you can split your file for upload:
    - **Maximum parts:** You can split your file into a maximum of 1000 parts.
    - **Minimum part size:** Each part must be at least 5MB in size. This is a temporary constraint and isn't currently configurable or discoverable. It will become an option when additional storage options are supported.
- **Initiating the upload:** When starting a multipart upload, you'll need to specify the total number of parts in your request. It reduce the requests for each part presigned URL when the client already knows the size.
- **Presigning part:** Presigned of a specific part URL for upload will not be supported. This will block unknown size upload using this API.

> - **Limited support:** Multipart uploads are not currently supported.

Reviewer: ?

Author: updated

# OpenAPI for S3 Like Multipart Upload Proposal

## Introduction

This proposal outlines the development of Open API endpoints to enable multipart upload capabilities similar to Amazon S3 multipart. The intention is to enhance the data uploading process for large files, ensuring efficient and reliable data transfer.

## Goal

- Implementing multi-part upload APIs with support for parallel upload optimizes data transfer for large files over varying network conditions using lakectl.

## Non-goals

- Support all underlying storage types as we can experiment with the one that currently support this feature.
- SDK to wrap the new API and support this capability.

## Features and Functionalities

To minimize the scope of this feature and understand the benefits before supporting all blockstores, this feature will be scoped to S3 blockstore and only if presigned capability is enabled.

> Initiation of Upload: The API will allow clients to initiate a multipart upload session, assigning a unique upload ID for subsequent operations. The client will require pass the number of parts it requires and the call will provide a set of URLs to upload each part.

Reviewer 3: ```suggestion
Initiation of Upload: The API will allow clients to initiate a multipart upload session, assigning a unique upload ID for subsequent operations. The client will require passing the number of parts it requires and the call will provide a set of URLs to upload each part.
```

        - staging
      operationId: abortMultipartUpload
      summary: Abort a multipart upload
      description: Aborts a multipart upload.
      responses:
        204:
          description: Multipart upload aborted

```

**Schemas used by operations**

```yaml
components:
  schemas:
    MultipartUpload:
      type: object
      properties:
        upload_id:
          type: string
        locations:
          type: array
          items:
            $ref: "#/components/schemas/StagingLocation"
      required:
        - upload_id
        
    UploadPart:
      type: object
      properties:
        part_number:
          type: integer
        etag:
          type: string
      required:
        - part_number
        - etag
        
    CompleteMultipartUpload:
      type: object
      properties:
        parts:
          type: array
          description: "List of uploaded parts, should be ordered by ascending part number"
          items:
            $ref: "#/components/schemas/UploadPart"
      required:
        - parts
```

### Support and discover

Presign support is a capability lakectl discover before switching to use presign for upload or download from lakeFS.

> The multipart upload support will be part of the storage capability add `multipart_upload_support` optional field that when set to `true` the user can perform multipart upload using the new API.

Reviewer 3: Why do we **actually** need this extra param?
Can't we rely on the block adapter type for now?

Author: It means that lakectl as a client will need to check:
1. s3 block adapter
2. presgined support enabled
3. lakefs version is above v1.x that include this feature

Reviewer 3: For 1,2 the client can get this information from the GetConfig API
Regarding no. 3 I think is not necessary, IMHO it's perfectly fine for the server to return 501 and for the client to handle it appropriately.
Adding this field will cause ambiguity, since it is a direct result of blockstore_type + pre_sign

Author: I took elad's suggestion and like a feature flag it represents the capability. if we like to support multipart through lakefs API we can.

The second issue related to try and fail does not help with client side code when you like to sync files in parallel while trying to understand one of you can use this feature. possible but not clean.
Also status code for unsupported operation in the logs vs clean 200 ok, I prefer the last.

The client will need to check also for `pre_sign_support` in order to request multipart upload with `presigned_parts`.
Currently as we only support multipart upload for persigned configuration, the server will turn off multipart upload support when presigned is false.

```yaml
    StorageConfig:
      type: object
      required:
        - blockstore_type
        - blockstore_namespace_example
        - blockstore_namespace_ValidityRegex
        - pre_sign_support
        - pre_sign_support_ui
        - import_support
        - import_validity_regex
      properties:
        blockstore_type:
          type: string
        blockstore_namespace_example:
          type: string
        blockstore_namespace_ValidityRegex:
          type: string
        default_namespace_prefix:
          type: string
        pre_sign_support:
          type: boolean
        pre_sign_support_ui:
          type: boolean
        import_support:
          type: boolean
        import_validity_regex:
          type: string
        pre_sign_support_parts:
          type: boolean
```

## Limitations

> - **S3 block adapter exclusive:** Uploading files in multiple parts (multipart upload) is only available when using the S3 block adapter. This feature isn't currently supported with other storage options. This exclude tha multipart upload support we provide by our S3 gateway where we provide implementation above each storage.

Reviewer 3: ```suggestion
- **S3 block adapter exclusive:** Uploading files in multiple parts (multipart upload) is only available when using the S3 block adapter. This feature isn't currently supported with other storage options. This exclude the multipart upload support we provide by our S3 gateway where we provide implementation above each storage.
```

- **Part size and count restrictions:** There are limits on how you can split your file for upload:
    - **Maximum parts:** You can split your file into a maximum of 10000 parts.
    - **Minimum part size:** Each part must be at least 5MB in size. This is a temporary constraint and isn't currently configurable or discoverable. It will become an option when additional storage options are supported.

> - **Initiating the upload:** When starting a multipart upload, you'll need to specify the total number of parts in your request. It reduce the requests for each part presigned URL when the client already knows the size.

Reviewer 3: ```suggestion
- **Initiating the upload:** When starting a multipart upload, you'll need to specify the total number of parts in your request. It will reduce the requests for each part's presigned URL when the client already knows the size.
```
